File: frontend/src/api/monitoring.ts
import { useQuery, useMutation } from "@tanstack/react-query";
import { api } from "./client";

export interface RunStats {
  total_runs: number;
  completed: number;
  failed: number;
  cancelled: number;
  running: number;
  pending: number;
  success_rate: number;
  avg_duration_seconds: number | null;
}

export interface DailyRunStat {
  date: string;
  completed: number;
  failed: number;
  total: number;
}

export interface CDCStats {
  total_jobs: number;
  idle: number;
  running: number;
  failed: number;
  paused: number;
  total_rows_synced: number;
}

export interface SystemStats {
  pipeline_stats: RunStats;
  cdc_stats: CDCStats;
  daily_runs: DailyRunStat[];
}

export function useMonitoringStats(days = 30) {
  return useQuery({
    queryKey: ["monitoring-stats", days],
    queryFn: async () => {
      const { data } = await api.get<SystemStats>(`/monitoring/stats?days=${days}`);
      return data;
    },
    refetchInterval: 10000,
  });
}

export function useExportLogs() {
  return useMutation({
    mutationFn: async (params: {
      format: "json" | "csv";
      days: number;
      include_pipeline_runs: boolean;
      include_cdc_logs: boolean;
    }) => {
      const { data } = await api.post("/monitoring/export", params, {
        responseType: "blob",
      });
      return { blob: data, format: params.format };
    },
  });
}

export function useExportToWebhook() {
  return useMutation({
    mutationFn: async (params: {
      url: string;
      secret?: string;
      days: number;
      include_pipeline_runs: boolean;
      include_cdc_logs: boolean;
    }) => {
      const { data } = await api.post("/monitoring/export/webhook", params);
      return data as { success: boolean; url: string; error?: string };
    },
  });
}

export function useTestWebhook() {
  return useMutation({
    mutationFn: async (params: { url: string; secret?: string }) => {
      const { data } = await api.post("/monitoring/webhook/test", params);
      return data as { success: boolean; url: string; error?: string };
    },
  });
}

File: backend/app/schemas/monitoring.py
"""Monitoring & observability schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class RunStats(BaseModel):
    total_runs: int = 0
    completed: int = 0
    failed: int = 0
    cancelled: int = 0
    running: int = 0
    pending: int = 0
    success_rate: float = 0.0
    avg_duration_seconds: Optional[float] = None


class DailyRunStat(BaseModel):
    date: str
    completed: int = 0
    failed: int = 0
    total: int = 0


class CDCStats(BaseModel):
    total_jobs: int = 0
    idle: int = 0
    running: int = 0
    failed: int = 0
    paused: int = 0
    total_rows_synced: int = 0


class SystemStats(BaseModel):
    pipeline_stats: RunStats
    cdc_stats: CDCStats
    daily_runs: list[DailyRunStat] = []


class LogExportRequest(BaseModel):
    format: str = "json"  # json, csv
    days: int = 30  # how many days of logs to include
    include_pipeline_runs: bool = True
    include_cdc_logs: bool = True


class WebhookExportRequest(BaseModel):
    url: str
    secret: str = ""
    days: int = 30
    include_pipeline_runs: bool = True
    include_cdc_logs: bool = True


class WebhookTestRequest(BaseModel):
    url: str
    secret: str = ""
